runtime/minor_gc.c

```diff
@@ -518,13 +519,97 @@ void caml_empty_minor_heap_domain_clear (struct domain* domain, void* unused)
+  int global_minor_heap_size = caml_global_minor_heap_limit - caml_global_minor_heap_start;
+  // Now using the number of participating domains, we calculate the new size
+  int new_global_minor_heap_size = participating_domains*Bsize_wsize(global_minor_heap_wsz_per_domain) + caml_global_minor_heap_start;
+
+  if( global_minor_heap_size != new_global_minor_heap_size ) {
+    caml_mem_decommit((char*)caml_global_minor_heap_start, global_minor_heap_size);
+    caml_mem_commit((char*)caml_global_minor_heap_start, new_global_minor_heap_size);
+
+    caml_global_minor_heap_limit = caml_global_minor_heap_start + new_global_minor_heap_size;
+  }
+
+  atomic_store_explicit(&caml_global_minor_heap_ptr,
+            caml_global_minor_heap_start,
+            memory_order_release);
+}
+
+int caml_replenish_minor_heap()
+{
+  caml_domain_state* domain_state = Caml_state;
+  uintnat cached_global_minor_heap_ptr;
+  uintnat new_alloc_ptr;
+
+  uintnat minor_buffer_wsize = global_minor_heap_wsz_per_domain >> 3;
+  uintnat minor_buffer_bsize = Bsize_wsize(minor_buffer_wsize);
+  uintnat requested_buffer_size;
+
+  CAML_EV_BEGIN(EV_MINOR_REPLENISH_BUFFER);
+
+  // Here we try to replenish our minor heap from the global minor heap
+  // This needs to be done in a loop because it could fail.
+  while (1) {
+    uintnat remaining_global_minor;
+
+    cached_global_minor_heap_ptr = atomic_load_explicit(&caml_global_minor_heap_ptr, memory_order_acquire);
+
+    remaining_global_minor = caml_global_minor_heap_limit - caml_global_minor_heap_ptr;
+
+    requested_buffer_size = remaining_global_minor < minor_buffer_bsize ? remaining_global_minor : minor_buffer_bsize;
+
+    CAMLassert(caml_global_minor_heap_start <= cached_global_minor_heap_ptr);
+    CAMLassert(cached_global_minor_heap_ptr <= caml_global_minor_heap_limit);
+
+    new_alloc_ptr = cached_global_minor_heap_ptr + requested_buffer_size;
+
+    if (new_alloc_ptr > caml_global_minor_heap_limit) {
+      CAML_EV_END(EV_MINOR_REPLENISH_BUFFER);
+      return 0;
+    }
+
+    /* CAS away and bump the allocation pointer: if it fails a domain likely
+       snatched the requested segment. Restart the loop, else success. */
+    if(atomic_compare_exchange_strong(&caml_global_minor_heap_ptr,
+				      &cached_global_minor_heap_ptr,
+				      new_alloc_ptr)) {
+      break;
+    };
+  }
+
+  // before we trample over our current buffer, record how much we've allocated
+  domain_state->stat_minor_words += Wsize_bsize (domain_state->young_end - domain_state->young_ptr);
+
+  // global_minor_heap_ptr is now our new minor heap for this domain
+
+  // I can't think of a situation where requested_buffer_size is not a word multiple
+  domain_state->minor_heap_wsz = Wsize_bsize(requested_buffer_size);
+
+  caml_update_young_limit(cached_global_minor_heap_ptr);
+
+  domain_state->young_start = (char*)cached_global_minor_heap_ptr;
+  domain_state->young_end = (char*)(cached_global_minor_heap_ptr + minor_buffer_bsize);
```

Comment:
Is this `minor_buffer_bsize` really correct? 
Aren't there code paths here where the `requested_buffer_size` is not equivalent to `minor_buffer_bsize`?